### ISSUE TITLE
Provide a way to find time until *closest* timer expires

### ISSUE THREAD

**Issue by jonhoo** [andrewjstone/ferris#3]
Imagine you have a set of timers, but you need to do a blocking operation (like a `read` system call). You'd like to set a timeout for the system call so that you can detect expired timers in a timely fashion, because the block could last for a long time. There isn't currently a way in `ferris` to get the time until the nearest timeout expires, which would be necessary to provide this kind of functionality. Instead, you have to set a timeout on the order of your finest resolution, which can result in lots of unnecessary wakeups.

**Comment by jonhoo** [comment]
Given #6, this may no longer be a feasible request...

**Comment by andrewjstone** [comment]
Yeah, this requires a different sort of structure like a priority queue. This just jogged my memory though. I actually had to implement this exact scenario in `amy` since the platform we are running on doesn't support timerfd. You can see the complete [pull request](https://github.com/andrewjstone/amy/pull/20)

Specifically I implemented a [timerheap](https://github.com/andrewjstone/amy/pull/20/commits/5565888b60a750362b58bd81289c9598ae82ed5c#diff-57ccfabeb872f7cef942f9d483861a83) and then [use it here](https://github.com/andrewjstone/amy/pull/20/commits/f18ff0cee7d65cd55a0629ef141211ae4f8f0301#diff-1a41e4dcb6cec5befecca71500e4d036R148)

Based on your needs, it's looking like I should pull this out into it's own crate so you can use it if you desire. Let me know if this interests you. As I feel it's pretty urgent I can probably get to it tonight or tomorrow.

**Comment by jonhoo** [comment]
The `timerheap` looks like pretty much the interface I'd want, though the high cost of `remove` (it has to walk all the timers) is pretty unfortunate. I'd recommend pulling it out anyway though, because then you could iterate on the implementation *behind* the interface :)

I moved to doing a linear scan of my keyspace (which is okay, because I know it's small) in https://github.com/mit-pdos/distributary/commit/07fc46d7ede39e2c571c31b615104550fcf535d4#diff-577e5739fd3edf2bd8bdf9d040800516R206, so there isn't any rush, but I would love to remove that code in favor of a simpler interface (assuming similar performance).

**Comment by andrewjstone** [comment]
Yeah, the remove cost is unfortunate :( I'm sure there is a way around it, using a secondary structure like is done in ferris. I should factor it out and fix this. Honestly, this conversation has been so interesting I'd like to do it right now. Unfortunately I must do a few hours of work that they are paying me for and hopefully I can get back to this later :)

**Comment by jonhoo** [comment]
Haha, yes, the story of my life. That's how I ended up most of my Rust libraries (hdrsample included!)

**Comment by andrewjstone** [comment]
Hey @jonhoo I added a new crate, [timer_heap](https://github.com/andrewjstone/timer_heap), that should do what you want. It specifically maintains an active set of keys so they can be removed in 0(1) time. This does add some overhead to expiry, as well as per-timer memory overhead, but I don't believe we can get around that if we want fast cancel/remove functionality. Please take a look and let me know what you think.
